### PR TITLE
[comet-parquet-exec] Use Parquet schema for scan instead of Spark schema

### PR DESCRIPTION
**Pull request by mbutrovich**
Currently we get the scan schema from the plan nodes scan schema, and then serialize that back to a Parquet schema, then parse that on the native side. This is lossy, particularly with timestamps. For example:

```
schema: message root {
  optional int64 _0 (TIMESTAMP(MILLIS,true));
  optional int64 _1 (TIMESTAMP(MICROS,true));
  optional int64 _2 (TIMESTAMP(MILLIS,true));
  optional int64 _3 (TIMESTAMP(MILLIS,false));
  optional int64 _4 (TIMESTAMP(MICROS,true));
  optional int64 _5 (TIMESTAMP(MICROS,false));
  optional int64 _6 (INTEGER(64,true));
}

dataSchema: message spark_schema {
  optional int96 _0;
  optional int96 _1;
  optional int96 _2;
  optional int64 _3 (TIMESTAMP(MICROS,false));
  optional int96 _4;
  optional int64 _5 (TIMESTAMP(MICROS,false));
  optional int64 _6;
}
```
The former is the original Parquet footer, the latter is what we get after going through Spark. We need the original to handle timestamps correctly in ParquetExec.

This PR extracts some code from elsewhere (CometParquetFileFormat, CometNativeScanExec) to read the footer from the Parquet file, and serialize the original metadata. We also now generate the projection vector on the Spark side because the required columns is in Spark schema format, so will not match the Parquet schema 1:1. On the native side, we now have to regenerate the required schema from the Parquet schema using the projection vector (converted to a DF ProjectionMask).

### CODE REVIEW

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 1]
```diff
@@ -2507,23 +2508,22 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
               })
             case _ =>
+              assert(false)
           }
 
-          val requiredSchemaParquet =
-            new SparkToParquetSchemaConverter(conf).convert(scan.requiredSchema)
-          val dataSchemaParquet =
-            new SparkToParquetSchemaConverter(conf).convert(scan.relation.dataSchema)
+          val projection_vector: Array[java.lang.Long] = scan.requiredSchema.fields.map(field => {
```
Comment (parthchandra) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:2522]: This change essentially means that any schema 'adaptation' made in `SparkToParquetSchemaConverter.convert` to support legacy timestamps and decimals will not be supported. But we will probably fail tests with incorrect results.
Also, Comet's Parquet file reader uses `CometParquetReadSupport.clipParquetSchema` to do similar conversion and it includes support for Parquet [field_id](https://github.com/apache/parquet-format/blob/c70281359087dfaee8bd43bed9748675f4aabe11/src/main/thrift/parquet.thrift#L473) which is desirable for delta sources like Iceberg.  
Basically a field_id, if present, identifies a field more precisely (in the event of field name changes) in a schema.

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 2]
```diff
@@ -3191,9 +3191,25 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
     partition.files.foreach(file => {
+      if (!schema_saved) {
+        // TODO: This code shouldn't be here, but for POC it's fine.
+        // Extract the schema and stash it.
+        val hadoopConf =
+          scan.relation.sparkSession.sessionState.newHadoopConfWithOptions(scan.relation.options)
+        val broadcastedHadoopConf =
+          sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+        val sharedConf = broadcastedHadoopConf.value.value
+        val footer = FooterReader.readFooter(sharedConf, file)
```
Comment (parthchandra) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:3208]: You're right. This can never be in production code. For one, this is expensive.

Comment (mbutrovich) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:3208]: In theory it's just replacing the call that currently takes place during the file_reader instantiation, but yeah I'm still curious if it's already cached somewhere. I see references within Spark to a `footersCache` so I'm curious to look for that as well.

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 1 above)
Comment (mbutrovich) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:2522]: The `SchemaConverter` seems like it could be handled in DF's `SchemaAdapter`. I'll look at `clipParquetSchema` as well, thanks!

Comment (mbutrovich) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:2522]: My concern is that...

1. Java side parses Parquet metadata, generates a Spark schema
2. Java side converts Spark schema to Arrow schema (following Comet conversion rules)
3. Serialize Arrow types, native side feeds this into ParquetExec as the data schema

...may yield different results than:

1. Java side serializes original Parquet metadata
2. Serialize schema message
3. Native side parses message, generates Arrow schema and feeds this into ParquetExec as the data schema

I guess I could exhaustively test this hypothesis with all types.

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 2 above)
Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:3208]: I didn't know about a footersCache in Spark. Could you share a link maybe?

Comment (mbutrovich) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:3208]: It's actually part of parquet-hadoop: https://github.com/apache/parquet-java/blob/1e04ec74060a04cb029b6e61cf1098c1aba1446e/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java#L525

Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:3208]: Ah, thanks. I don't think we ever travel this path.
